### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by WenzhengZhang**
Hi @karlstratos , could you do a code review for me? Thanks!

### CODE REVIEW

File: zeshel/retrieval/data.py [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+    # ids and token masks
+    def get_entity_window(en_page_id):
+        page_id = en_page_id
+        en_title = doc[page_id]['title']
+        en_text = doc[page_id]['text']
+        max_len = len_max - 2  # cls , sep
+        ENT = '[unused2]'
+        en_title_tokens = tokenizer.tokenize(en_title)
+        en_text_tokens = tokenizer.tokenize(en_text)
+        window = (en_title_tokens + [ENT] + en_text_tokens)[:max_len]
```
Comment (karlstratos) [inline zeshel/retrieval/data.py:206]: No need for change, but just a comment that en_text_tokens already contains en_title_tokens (I think) so this formulation is redundant, and maybe skipping the title will improve performance because we see more text. Something to check in the future.

File: zeshel/retrieval/data.py [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+            model.module.evaluate_on = True
+            model.module.candidates_embeds = all_candidates_embeds
+        else:
+            model.evaluate_on = True
+            model.candidates_embeds = all_candidates_embeds
+    hard_indices = []
+    all_candidates_probs = []
+    with torch.no_grad():
+        for i, batch in tqdm(enumerate(mention_loader)):
+            if not too_large:  # entity embeddings are too large for memory
```
Comment (karlstratos) [inline zeshel/retrieval/data.py:336]: Again just a minor comment (no need for change), but it's perhaps a better practice to pass an argument batch_size_small=-1 and change this to "if batch_size_small < 0:", so that we can process more than 1 entity.

File: zeshel/retrieval/model.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+        Attend over ys with query xs to obtain weights, then apply weights to
+        values (ys if yalues is None)
+        Args:
+            xs: B x query_len x dim (queries)
+            ys: B x key_len x dim (keys)
+            mask_ys: B x key_len (mask)
+            values: B x value_len x dim (values); if None, default to ys
+        """
+        if self.attn == 'cosine':
+            l1 = self.cosine(xs, ys).unsqueeze(self.dim - 1)
```
Comment (karlstratos) [inline zeshel/retrieval/model.py]: Oops, cosine similarity is not right here. The reason is the values are bounded in [-1, 1], so doing a softmax over them can never yield a peaky attention distribution. You have to do really weird tricks like adding extra parameters just to rescale (see Section 4.1 of Gillick et al.)! I'd suggest removing this option if you didn't find it helpful.

File: zeshel/retrieval/model.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        bsz, l_x, d = ys.size()
+        attended_embeds = self.attention(xs, ys,
+                                         mask_ys=mask_ys,
+                                         values=values)  # (B,CT_y,d)
+        scores = (attended_embeds * xs).sum(-1).reshape(
+            bsz, C, l_y).sum(-1)
+        return scores
+
+
+class UnifiedModel(nn.Module):
```
Comment (karlstratos) [inline zeshel/retrieval/model.py]: Just a suggestion for a more explicit name, maybe "UnifiedRetriever"? We have another model for reranking. In general it's good to make names as detailed as possible within clarity.

**Review by karlstratos — COMMENTED** [review]
Looking good overall.
